### DISCUSSION TITLE
getting_started/first_3d_game/09.adding_animations

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#76] [User-contributed notes]
# getting_started/first_3d_game/09.adding_animations

In this final lesson, we'll use Godot's built-in animation tools to make our characters float and flap. You'll learn to design animations in the editor and use code to make your game feel alive. im...

https://docs.godotengine.org/en/stable/getting_started/first_3d_game/09.adding_animations.html

**Comment by Theodophilus** [comment]
Hey folks, if you run into an issue adding an additional Animation after the initial Player Animation, because the 'Animation' button in the dock isn't working/greyed out? I'm using 4.2.1, and if you Un-Pin the dock in 'player.tscn', it unlocks the 'Animation' button if you select an 'AnimationPlayer' node in 'mob.tscn'. I hope this helps!

![Notice how the 'Animation' Button is Greyed Out?](https://raw.githubusercontent.com/Theodophilus/CIS126VM/master/Godot-Tut-4.3-Error.png)

**Comment by Wadahell** [comment]
Animation > Manage Animations....
![image](https://github.com/user-attachments/assets/04d6f883-b9fd-480f-bbe6-4e310839e87d)

**Reply by Jinfei-Wang** [reply]
Exactly what I need

**Comment by BryceDMonaco** [comment]
I misread the tutorial and accidentally applied the animations to the transforms of the root Player and Mob nodes of their respective scenes, which locked all of their positions to the center of the screen and made the mobs just spawn at the same spot as the player.

If anyone else makes this mistake, delete your animation tracks and redo them with `Player(or Mob)/Pivot/Character` as the target. It looks like there is currently an [open issue](https://github.com/godotengine/godot-proposals/issues/4847) to make switching animation targets easier, hopefully that gets picked up in the future.

**Reply by renatocn86** [reply]
I had a similar problem. Enemies spawned in the same position as the player, but only when 'Autoplay on load' was activated in the mob's animation (I don't know why). 

I found two solutions without redoing the animation:
1- instead of copying the 'float' animation, I saved the animation library and loaded it into the animation manager of the mob's AnimationPlayer.
2- ─ the lazy solution ─ I copied the 'AnimationPlayer' node from `Player/Pivot/Character` to `Mob/Pivot/Character`.

**Comment by epicbattle1** [comment]
why do animations cause input delay?

**Reply by epicbattle1** [reply]
specifically, jumping is not consistent

**Reply by epicbattle1** [reply]
oh ok, the character does a very little hop for every other jump

**Reply by Calinou** [reply]
The running animation still plays while airborne, which is why it looks like the jump height varies when you jump when moving. However, it's actually always the same (and therefore, the jump behavior is predictable with regards to collision).

**Reply by DSchlossarek** [reply]
You can fix that by creating a new animation for the player and leave it empty. I called my new animation "idle". In the playsers _physics_process-function, add:

	# Play animations when on ground, but not when in air
	if not is_on_floor():
		#print("in air")
		$Pivot/Character/AnimationPlayer.play("idle")
	else:
		#print("on ground")
		$Pivot/Character/AnimationPlayer.play("float")

**Comment by ryanttb** [comment]
I understand what this is doing, generally, but maybe break down the specific values and explain how each is affecting the different components?

`$Pivot.rotation.x = PI / 6 * velocity.y / jump_impulse`

**Reply by Nedus** [reply]
So we can do it here!

`$Pivot.rotation.x = PI / 6 * (velocity.y / jump_impulse)`

- $Pivot.rotation.x is the "pitch" of the player. We want the nose to go up when player goes up (jump) and point down when it falls (by gravity)
- PI / 6 = 30° is what we decided the maximum pitch should be
- (velocity.y / jump_impulse) = (actual player vertical velocity / 20). 
   -> When player jumps the (velocity.y / jump_impulse)  is (20 / 20) = +1
   -> Then player start to fall and velocity.y becomes increasingly negative, then it can be (-15 / 20) = -0.75, and then (-20 / 20) = -1 or even (-30 / 20) = -1.5

Now you have a proportional value based on how fast the character is moving vertically.
So, every frame, the pitch of the player nose will gradually be set following the vertical velocity of the player, pitching gradually up to 30° when the player jumps (30° * +1 = +30°) and pitching gradually down when the player falls (30° * -1 = -30°).

There will be no pitching when the player is not jumping

**Comment by YoshiDanyo** [comment]
In the .7 seconds step
The tutorial does not indicate which properties and which value put in them

**Reply by Calinou** [reply]
It's telling you to move the keyframe for the **Position** property, which is part of any Node2D.

**Comment by jeffreysanglin** [comment]
I made a pause screen by combining with other tutorials!
They key was creating a new PauseNode that processed during game pause and took an input. It took a bit to get the pause screen to play nicely with the restart screen, but some simple logic statements were the fix. 🏆

**Reply by jeffreysanglin** [reply]
Btw I didn't even realize a pause mode was built in. Coming from monogame, that was super exciting. More tutorials on the kinds of native features would be super helpful. (Maybe they're there, and I just haven't gotten to them yet)

**Comment by DinkledDoo** [comment]
For anyone that dislikes using "PI" for the "Pivot.rotation", dont forget that you can always use "deg_to_rad"
e.g - $Pivot.rotation.x = deg_to_rad(180 / 6 * velocity.y / jump_impulse)

**Comment by rjbray915** [comment]
I noticed that when jumping and moving at the same time, the Player looks like it does another jump mid-air. I modified mine to only increase SpeedScale while on the ground:
```
if (IsOnFloor())
    GetNode<AnimationPlayer>("AnimationPlayer").SpeedScale = 4;
else
    GetNode<AnimationPlayer>("AnimationPlayer").SpeedScale = 1;
```

**Reply by andrefuture81** [reply]
But won't that mean that in your "else" block (when the player is airborne) the float animation will still play, just at 1x speed? 

I had the same issue and set the speed scale to 0 when the player is airborne which got rid of that ugly double jump looking thing.

**Comment by ghillahill** [comment]
I don't know what it is but the Player Character stop rotate after setting the animation player.

Using Godot 4.4

**Reply by RenionDev** [reply]
It seems that this is **a common issue in Godot 4.x**
What I found to be the issue is that `AnimationPlayer` overwrites the properties of the `Pivot` if you follow the step-by-step guide. This is problematic since it overwrites the way you control the player's direction through the following line:
GDScrips :
```$Pivot.basis = Basis.looking_at(direction)```
C# :
```GetNode<Node3D>("Pivot").Basis = Basis.LookingAt(direction);```
The workaround I found is to connect the animations to the `Character` Node instead, which doesn't overwrite the rotation set to the pivot through the movement inputs.

This seems to be working fine, although you might find the animation a bit clunky.

**Comment by vezocoding** [comment]
Very nice tutorial! But I have one question: I do not see the shadows from the example gif on top of this chapter when playing. Can someone help?

**Reply by Calinou** [reply]
Try adjusting the **Shadow Max Distance** property in the DirectionalLight3D node in the inspector. If it's too low, shadows will cut off early. If it's too high, shadows will appear blurry which may hide their actual appearance.

**Comment by liebhaber-von-dunkelheit** [comment]
So for this last part here where we add:

(
func initialize(start_position, player_position):
	#...
	$AnimationPlayer.speed_scale = random_speed / min_speed
)

The GDScript warning system warns about dividing integers, and namely that any decimal component of the answer will be discarded. since we're calculating somewhere between 18/10 and 10/10, discarding of the decimal component renders this operation basically useless given the context of coming out between 1.8 and 1.0; We're either choosing between 2 and 1, or just always choosing 1 if it doesn't bother to round up.

What I *think* a good solution is (and definitely please correct me if I'm wrong) to this issue is explicitly telling the script that you want to use each of these variables (random_speed and min_speed) as floats instead of integers to preserve the resulting decimal; so a more effective way to write it would be this:

(
func initialize(start_position, player_position):
	#...
	$AnimationPlayer.speed_scale = float(random_speed) / float(min_speed)
)

I honestly can barely tell the difference either way, but this way makes the editor (and my brain) happier at least.

**Comment by Pinaz993** [comment]
```$Pivot.rotation.x = PI / 6 * velocity.y / jump_impulse```
I didn't want to put the above statement at the end of `_physics_process()` because it felt out of place when I'd gone through so much trouble to put all the position logic in one place. So, I decided to put it under `if not is_on_floor()`. This caused a small issue: the player would land on their nose and remain in that X rotation until the jumped again. I did some rewriting, and my result now looks like this:
```gdscript
# Determine vertical velocity.
	if not is_on_floor(): # We are not on the ground and thus cannot jump and should fall.
		target_velocity.y = target_velocity.y - (fall_acceleration * delta)
		# Update rotation so that it looks like the player purposefully jumps .
		$Pivot.rotation.x = PI / 6 * target_velocity.y / jump_impulse
	# End if. We are on the ground. Has jump been recently pressed? 
	elif Input.is_action_just_pressed("jump"): # Recently pressed because jump buffering feels good.
		target_velocity.y = jump_impulse # Set the vertical velocity to rising.
	else: # We aren't jumping or in the air
		$Pivot.rotation.x = 0 # Zero out the x so that the player isn't standing on its nose.
```

**Comment by MiaoooOwO** [comment]


**Reply by Pinaz993** [reply]
Thank you! Also, this seems to be the first time that I've been replied to in a language other than English. Very interesting indeed.

**Comment by Pinaz993** [comment]
Indeed, that is a distinct possibility, considering English's position as a
lingua franca for large segments of the internet.

I really do enjoy the international flavor of various online communities.
The diversity of humankind is truely something to be cherished and
celebrated.

**Comment by somebrownnerd** [comment]
I'm using 4.6. When I load into my game the character preforms the animation once, then hangs on the last position of the animation. The animation is not looping. Has anyone had a similar issue?

**Reply by Calinou** [reply]
Check the looping icon at the right of the animation editor, after selecting your animation in the dropdown:

<img width="243" height="103" alt="image" src="https://github.com/user-attachments/assets/fdac42be-7402-4bc7-9128-d5804241e872" />

Note that there are two "enabled" modes, wrap (the default) and clamp (with a different icon than when looping is disabled). Click on the icon to cycle through the modes.

**Comment by eatthischair** [comment]
two things:
on the docs, the animation keyframe node is listed under Player, but when adding position or rotation keyframe to Player (as mentioned above), they reset the position of the Player every 1.2 seconds
Mine (which works fine) has the animations listed under Character (right under Filter Tracks). Maybe the pictures are from an old version

Secondly: I wasn't able to copy the animation as when I open up the anim pasted from clipboard in the Mob scene, the animation node (under filter tracks) lists as Pivot/Character, instead of just Character as it is in the Player scene.
No big for this project, but curious why this is for future projects
